### PR TITLE
ASIC internal temperature sensors support

### PR DESCRIPTION
**Pull request by santhosh-kt**
**What I did**
1) ASICs have multiple internal thermal sensors.
2) With the help of configuration, a poller is introduced in Orchagent that will periodically retrieve values of these sensors with the help of SAI APIs (opencomputeproject/SAI#880).
3) These retrieved values are being populated to the state DB (In "ASIC_TEMPERATURE_INFO" table).

**Why I did it**
As part of ASIC Thermal Monitoring HLD.

**How I verified it**
Tested out in below platforms.
[asic_thermal_log.s6100.txt](https://github.com/Azure/sonic-swss/files/5582296/asic_thermal_log.s6100.txt)
[asic_thermal_log.z9264.txt](https://github.com/Azure/sonic-swss/files/5582297/asic_thermal_log.z9264.txt)
[asic_thermal_log.s5232.txt](https://github.com/Azure/sonic-swss/files/5582298/asic_thermal_log.s5232.txt)
[asic_thermal_log.s6000.txt](https://github.com/Azure/sonic-swss/files/5582295/asic_thermal_log.s6000.txt)

### Configuration Sample(config_db.json)
1) To enable the poller:
```
  "ASIC_SENSORS": {
          "ASIC_SENSORS_POLLER_STATUS": {
                 "admin_status": "enable"                              
          }
   }
```
2) To disable the poller:
```
  "ASIC_SENSORS": {
          "ASIC_SENSORS_POLLER_STATUS": {
                 "admin_status": "disable"                              
          }
   }
```
3) To change the poller interval(Default time interval = 60 seconds):
```
  "ASIC_SENSORS": {
          "ASIC_SENSORS_POLLER_STATUS": {
                 "admin_status": "enable"                              
          },
          "ASIC_SENSORS_POLLER_INTERVAL": {
                 "interval": "30"
          }
   }
```

### Test cases and their observations

Steps | Test Case   scenarios | Result
-- | -- | --
(1) | Enabling the poller with default time interval in   config-db.json and do “config reload” | Poller populated the “ASIC_TEMPERATURE_INFO” table in the state DB and updated the table every default interval.
(2) | Disabling the poller in config-db.json and do “config   reload” | Poller disabled and the thermal sensors readings in   “ASIC_TEMPERATURE_INFO” table are reset to 0
(3) | Enabling the poller and setting the interval to “30” in   config-db.json and do “config reload” | Poller populated the “ASIC_TEMPERATURE_INFO” table in the state DB and updated the table every 30 seconds.
(4) | Disabling the poller by removing the configs from the   config-db.json and do “config reload” | Poller disabled and the thermal sensors readings in   “ASIC_TEMPERATURE_INFO” table are reset to 0
(5) | Repeating the steps (1), (2), (3), and (4) but with   “reboot” instead of “config reload” | Same results as above
(6) | Repeating the steps (1), (2), (3), and (4) but with   “fast-reboot” instead of “config reload” | Same results as above
(7) | Downgrading the SONiC with config enabling the poller | Configuration got neglected since not supported in the older   version
(8) | Upgrading the SONiC with config enabling the poller | Poller ran and updated the “ASIC_TEMPERATURE_INFO” table in the state DB and updated the table every default interval.

### Sample RedisDB ASIC Sensors Table Output(S6100)
Output: 443 => 44.3 degC
```
root@sonic:~# redis-cli -n 6 hgetall "ASIC_TEMPERATURE_INFO"
1) "temperature_0"
2) "443"
3) "temperature_1"
4) "452"
5) "temperature_2"
6) "457"
7) "temperature_3"
8) "447"
9) "temperature_4"
10) "433"
11) "temperature_5"
12) "457"
13) "temperature_6"
14) "423"
15) "temperature_7"
16) "443"
17) "maximum_temperature"
18) "477"
19) "average_temperature"
20) "451"
root@sonic:~#
```
**Details if related**
HLD Location: https://github.com/Azure/SONiC/blob/master/doc/asic_thermal_monitoring_hld.md

### CODE REVIEW

**Comment by santhosh-kt** [comment]
vs build failed due to missing dependency, Azure/sonic-swss-common#419

**Comment by jleveque** [comment]
https://github.com/Azure/sonic-swss-common/pull/419 has merged. Submodule update PR is here: https://github.com/Azure/sonic-buildimage/pull/6042

**Comment by santhosh-kt** [comment]
retest this please

**Comment by santhosh-kt** [comment]
retest this please

**Comment by santhosh-kt** [comment]
retest LGTM please

**Comment by santhosh-kt** [comment]
retest this please

**Comment by santhosh-kt** [comment]
retest this please

**Comment by padmanarayana** [comment]
@jleveque @keboliu  : please review

**Comment by jleveque** [comment]
LGTM. @kcudnik to review, as well.

File: orchagent/switchorch.cpp [hunk 1]
```diff
@@ -145,6 +228,27 @@ void SwitchOrch::doTask(Consumer &consumer)
     }
 }
 
+void SwitchOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    const string & db_name = consumer.getDbName();
+    const string & table_name = consumer.getTableName();
+
+    if ((db_name == "APPL_DB") && (table_name == APP_SWITCH_TABLE_NAME ))
```
Comment (kcudnik) [inline orchagent/switchorch.cpp:237]: dont use direct names APPL_DB, we have defines for those in headers

File: orchagent/switchorch.cpp [hunk 2]
```diff
@@ -33,17 +33,100 @@ const map<string, sai_packet_action_t> packet_action_map =
...
+                        m_sensorsPollerIntervalChanged = true;
+                    }
+                    else
+                    {
+                        SWSS_LOG_INFO("ASIC sensors : poller interval unchanged : %d seconds",m_sensorsPollerInterval);
+                    }
+                }
+                else
+                {
+                    SWSS_LOG_ERROR("ASIC sensors : unsupported field");
```
Comment (kcudnik) [inline orchagent/switchorch.cpp:112]: should you log also on which component this field is not supported ?, same on 117

File: orchagent/switchorch.cpp (hunk 1 above)
Comment (kcudnik) [reply orchagent/switchorch.cpp:237]: same, 241

File: orchagent/switchorch.cpp [hunk 3]
```diff
@@ -209,6 +313,175 @@ bool SwitchOrch::setAgingFDB(uint32_t sec)
...
+        }
+        else
+        {
+            SWSS_LOG_ERROR("ASIC sensors : failed to get SAI_SWITCH_ATTR_TEMP_LIST: %d", status);
+        }
+    }
+
+    if (m_sensorsMaxTempSupported)
+    {
+        memset(&attr, 0, sizeof(attr));
```
Comment (kcudnik) [inline orchagent/switchorch.cpp:468]: why attr is clear if not used?

Comment (kcudnik) [reply orchagent/switchorch.cpp:468]: same on 477

File: orchagent/switchorch.cpp (hunk 1 above)
Comment (prsunny) [reply orchagent/switchorch.cpp:237]: But you can just check for table_name right?

**Review by prsunny — COMMENTED** [review]
Should we add VS tests?

File: orchagent/switchorch.cpp (hunk 2 above)
Comment (santhosh-kt) [reply orchagent/switchorch.cpp:112]: will add the changes

File: orchagent/switchorch.cpp (hunk 1 above)
Comment (santhosh-kt) [reply orchagent/switchorch.cpp]: Added db check similar to pfcwdorch.cpp 
Now removed the db check.

**Comment by santhosh-kt** [comment]
retest this please

**Comment by padmanarayana** [comment]
@kcudnik , @prsunny : can you please review the changes ?

**Review by kcudnik — APPROVED** [review]
please wait for others to approve

**Comment by santhosh-kt** [comment]
@prsunny @keboliu : Please re-review the changes,

**Review by prsunny — APPROVED** [review]
lgtm..@keboliu

**Comment by santhosh-kt** [comment]
@keboliu  Please review.

**Review by keboliu — APPROVED** [review]
LGTM

**Comment by jleveque** [comment]
@santhosh-kt: All reviewers have signed off and https://github.com/Azure/sonic-swss-common/pull/419 has been merged and the submodule has been updated to include that commit. Are we good to merge this now?

**Comment by santhosh-kt** [comment]
@jleveque : Please proceed to merge and update the submodule.

**Comment by santhosh-kt** [comment]
@jleveque : This PR is yet to get update in the submodule.

**Comment by jleveque** [comment]
@santhosh-kt: Please feel free to open a PR to update the sonic-swss submodule in sonic-buildimage.

**Pull request merged**